lightning/src/ln/chan_utils.rs

```diff
@@ -550,11 +556,441 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, contest_del
+	/// This allows the caller to match the HTLC output index with the auxiliary data.
+	/// This auxiliary data is not stored in this object.
+	pub fn new_with_auxiliary_htlc_data<T: Copy>(commitment_number: u64, to_broadcaster_value_sat: u64, to_countersignatory_value_sat: u64, keys: TxCreationKeys, feerate_per_kw: u32, htlcs_with_aux: Vec<(HTLCOutputInCommitment, T)>, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<secp256k1::All>) -> (CommitmentTransaction, Vec<(HTLCOutputInCommitment, T)>) {
+		// Sort outputs and populate output indices while keeping track of the auxiliary data
+		let mut txouts = Self::do_build_outputs(&keys, to_broadcaster_value_sat, to_countersignatory_value_sat, &htlcs_with_aux, channel_parameters, &secp_ctx).unwrap();
+		let mut result_htlcs_with_aux = Vec::new();
+		let mut htlcs = Vec::new();
+		for (idx, mut out) in txouts.drain(..).enumerate() {
+			if let Some(mut htlc) = (out.1).1.take() {
+				htlc.0.transaction_output_index = Some(idx as u32);
+				result_htlcs_with_aux.push((htlc.0.clone(), htlc.1));
+				htlcs.push(htlc.0);
+			}
+		}
+
+		let info = CommitmentTransaction {
+			commitment_number,
+			to_broadcaster_value_sat,
+			to_countersignatory_value_sat,
+			feerate_per_kw,
+			htlcs,
+			keys,
+		};
+		(info, result_htlcs_with_aux)
+	}
+
+	pub(crate) fn txid<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Txid {
+		let tx = self.build(channel_parameters, secp_ctx).unwrap().0;
+		tx.txid()
+	}
+
+	/// Build the Bitcoin transaction
+	pub fn build<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<(bitcoin::Transaction, Vec<HTLCOutputInCommitment>, Vec<Script>), ()> {
+		let (obscured_commitment_transaction_number, txins) = self.build_inputs(channel_parameters);
+
+		let mut txouts = self.build_outputs(channel_parameters, secp_ctx)?;
+
+		let mut outputs = Vec::with_capacity(txouts.len());
+		let mut scripts = Vec::with_capacity(txouts.len());
+		let mut htlcs = Vec::new();
+		for (idx, mut out) in txouts.drain(..).enumerate() {
+			outputs.push(out.0);
+			scripts.push((out.1).0.clone());
+			if let Some(mut htlc) = (out.1).1.take() {
+				htlc.transaction_output_index = Some(idx as u32);
+				htlcs.push(htlc);
+			}
+		}
+
+		Ok((
+			Transaction {
+				version: 2,
+				lock_time: ((0x20 as u32) << 8 * 3) | ((obscured_commitment_transaction_number & 0xffffffu64) as u32),
+				input: txins,
+				output: outputs,
+			},
+			htlcs,
+			scripts,
+		))
+	}
+
+	fn build_outputs<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_parameters: &DirectedChannelTransactionParameters, secp_ctx: &Secp256k1<T>) -> Result<Vec<(TxOut, (Script, Option<HTLCOutputInCommitment>))>, ()> {
+		let htlcs = self.htlcs.iter().map(|h| (h.clone(), ())).collect();
```

Comment:
This seems pretty redundant - are there callers where we really need clones of `HTLCOutputInCommitment` or can we just let clients access the already-newly-pub `htlcs` field in the struct?

Reply:
this was mostly because I didn't want to duplicate `do_build_outputs` for the two cases: initial sort of `HTLCSource` aux data, and later building of the transaction.  I cleaned this up in https://github.com/rust-bitcoin/rust-lightning/pull/742/commits/3c35f8d3afe3775c027efa0e3e82c0c7ce3df6eb and https://github.com/rust-bitcoin/rust-lightning/pull/742/commits/b2ce5e8012dbb5aa40244de09cdfe9a3fa49965f - let me know what you think.